Modules/Campaign/Http/Services/CampaignService.php

```diff
@@ -1,152 +1,304 @@

-                'data' => ""

-            ];

-        }

-

-        return [

-            'res' => true,

-            'msg' => '',

-            'data' => $campaign

-        ];

-    }

-

-    /**

-     * Remove the specified campaign from storage.

-     *

-     * @param string $campaignKey

-     * @return array

-     */

-    public function delete(string $campaignKey): array

-    {

-        $campaign = Campaign::where('campaign_key', $campaignKey)->first();

-

-        // return error if no campaign found

-        if (!$campaign) {

-            return [

-                'res' => false,

-                'msg' => 'Campaign not found !',

-                'data' => ""

-            ];

-        }

-

-        $campaign->delete();

-

-        return [

-            'res' => true,

-            'msg' => 'Campaign successfully deleted',

-            'data' => ""

-        ];

-    }

-

-}

+<?php
+
+
+
+namespace Modules\Campaign\Http\Services;
+
+
+
+use Illuminate\Support\Str;
+
+use Modules\User\Entities\User;
+
+use Modules\Campaign\Entities\Campaign;
+
+
+
+
+
+class CampaignService
+
+{
+
+    protected Campaign $campaign;
+
+
+
+    protected User $user;
+
+
+
+    /**
+
+     * Save a new campaign
+
+     *
+
+     * @param array $requestData
+
+     * @return array
+
+     */
+
+    public function store(array $requestData): array
+
+    {
+
+
+
+        //set request data with authenticated user id.
+
+        $requestData["status"] = Campaign :: STATUS_DRAFT;
+
+        $campaign = $this->createCampaign($requestData);
+
+        $response = [
+
+            'res' => true,
+
+            'msg' => 'Your campaign created successfully',
+
+            'data' => $campaign
+
+        ];
+
+
+
+        return $response;
+
+    }
+
+
+
+    /**
+
+     * Create a new campaign
+
+     *
+
+     * @param array $campaignData
+
+     * @return Campaign
+
+     */
+
+    public function createCampaign(array $campaignData): Campaign
+
+    {
+
+
+
+        //set campaign data
+
+        $campaignData["campaign_key"] = 'bmc_' . Str::lower(Str::random(10));
+
+
+
+        //create campaign
+
+        $campaign = new Campaign();
+
+        $campaign->fill($campaignData);
+
+        $campaign->save();
+
+
+
+        return $campaign;
+
+    }
+
+
+
+    /**
+
+     * Get a listing of the campaigns
+
+     *
+
+     * @param $requestData
+
+     * @return array
+
+     */
+
+    public function getCampaigns($requestData): array
+
+    {
+
+
+
+        $allCampaignsCount = Campaign::where('user_id', $requestData->user_id)->count();
+
+        $draftCampaignsCount = Campaign::where('user_id', $requestData->user_id)->where('status', 'draft')->count();
+
+        $scheduledCampaignsCount = Campaign::where('user_id', $requestData->user_id)->where('status', 'schedule')->count();
+
+        $completedCampaignsCount = Campaign::where('user_id', $requestData->user_id)->where('status', 'completed')->count();
```

Comment:
@arijit1111 why are we querying 4 time for just the counts? Can't we query at once and get the counts based on status?
Like this:
$campaignCounts = Campaign::selectRaw('status, count(*) as count')
    ->where('user_id', $requestData->user_id)
    ->groupBy('status')
    ->get()
    ->pluck('count', 'status');

$allCampaignsCount = $campaignCounts->sum();
$draftCampaignsCount = $campaignCounts->get('draft', 0);
$scheduledCampaignsCount = $campaignCounts->get('schedule', 0);
$completedCampaignsCount = $campaignCounts->get('completed', 0);

You can test the query, but we should send the request to DB at once not four times.